StatefulSet Support for Updating Volume Claim Template

huww98 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Implementation for https://github.com/kubernetes/enhancements/issues/4650

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
https://github.com/kubernetes/enhancements/issues/4650

#### Special notes for your reviewer:

WIP: preview of API change to statefulset spec

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Adds a feature gate UpdateVolumeClaimTemplate, which allow limited update to VolumeClaimTemplates of StatefulSet, and reconcile the updates with PVCs.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4650
```

huww98 committed:
check each immutable statefulset field seperately

Now we have more mutable fields than immutable ones (6 vs 5), and more will become mutable soon.
Avoids the deep copy. Make the error message more consistent and specific.

huww98 committed:
allow update size to STS volumeClaimTemplates

huww98 committed:
add feature gate UpdateVolumeClaimTemplate

huww98 committed:
add volumeClaimTemplates to STS history

only if it is updated at least once since the UpdateVolumeClaimTemplate is enabled.

huww98 committed:
add field VolumeClaimUpdatePolicy to StatefulSet

Also added defaults, validation. The field is dropped if feature gate is not enabled as recommended in
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version

And run the code generations:

./hack/update-codegen.sh protobuf swagger defaults conversions openapi applyconfigs protobindings
./hack/update-openapi-spec.sh
UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test k8s.io/api -run //HEAD

huww98 commented:
/sig storage
/test pull-kubernetes-verify-lint
/test pull-kubernetes-verify
/test pull-kubernetes-dependencies

googs1025 commented on pkg/apis/apps/types.go:
```diff
@@ -239,6 +255,13 @@ type StatefulSetSpec struct {
 	// +optional
 	Ordinals *StatefulSetOrdinals
+
+	// volumeClaimUpdatePolicy indicates when PersistentVolumeClaims should be
```
nit: VolumeClaimUpdatePolicy

huww98 replied:
I'm not sure. This is not consistent. e.g. `ordinals`, `revisionHistoryLimit` above also starts with lower case.

googs1025 commented on pkg/apis/apps/validation/validation.go:
```diff
@@ -88,6 +90,18 @@ func ValidatePersistentVolumeClaimRetentionPolicy(policy *apps.StatefulSetPersis
+func ValidateVolumeClaimUpdatePolicy(policy apps.StatefulSetVolumeClaimUpdatePolicyType, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	switch policy {
+	case "":
```
Does this policy field have a default value? If it is empty, is it necessary to check for errors?

huww98 replied:
I'm following this guide https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version
> 5. In validation, validate the field if present

If the feature gate is not enabled, it is empty. Or it will have default value.

huww98 committed:
controller/statefulset: collect reversion info into updateContext

These info is used in several places, collect them to reduce number of parameters.

huww98 committed:
controller/statefulset: firstUnhealthy{Pod => Ordinal}

Should allow we delay the creation of new pods

huww98 committed:
controller/statefulset: set correct revision for creating PVCs

Enabled under feature gate.
Delay creation of new pods to processReplica. Avoid creation of unnessesary pod objects.

huww98 committed:
controller/statefulset: wait for resizing to finish during rollout

huww98 committed:
controller/statefulset: create PVC with server-side apply

huww98 committed:
controller/statefulset: update PVC before creating new Pod

huww98 committed:
controller/statefulset: only update PVC if pod is unchanged

huww98 committed:
controller/statefulset: remove duplicate history.SortControllerRevisions()

huww98 committed:
controller/history: use maps.Clone()

huww98 committed:
controller/statefulset/truncateHistory: use set to replace map

huww98 committed:
controller/statefulset/updateStatus: use updateContext and simplify

huww98 committed:
controller/statefulset: record pod up-to-date revision names

huww98 committed:
add patch PVC permission to statefulset-controller

huww98 committed:
controller/statefulset: handle PVC update event

To kick rollout process moving.

ibihim commented:
/triage accepted

yxxhero commented:
/reopen

yxxhero commented:
@huww98 looking forward this PR. it will fix the resize pvc issue when use helm to upgrade sts.

huww98 commented:
/reopen
/remove-lifecycle rotten

eakman-datadog commented on pkg/controller/statefulset/stateful_pod_control.go:
```diff
@@ -363,3 +424,37 @@ func (spc *StatefulPodControl) createPersistentVolumeClaims(set *apps.StatefulSe
+		}
+		_, err = spc.objectMgr.ApplyClaim(ctx, &template)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to apply PVC %s: %s", claim.Name, err))
```
In the case where both `apierrors.IsNotFound(err)` and `create` are `true`, what is `claim.Name` going to be? Does it make sense to grab the name from the template?

eakman-datadog replied:
What I mean is on line 465 where we append the error to `errs`, we're using `claim.Name` in the error message. `claim` comes from line 435 where we call `claim, err := spc.objectMgr.GetClaim(template.Namespace, template.Name)`. If that call yields an a not found error, then I'm not sure what `claim` would be/whether it would be proper to use `claim.Name` in the error message. What I'm suggesting instead is to use `template.Name` which is like you said generated using `getPersistentVolumeClaimName`. Does that make sense? Am I misunderstanding?

eakman-datadog replied:
NP!

eakman-datadog reviewed (commented):
Look great so far!

eakman-datadog commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -495,6 +490,28 @@ func isHealthy(pod *v1.Pod) bool {
+}
+
+// prepareClaimApplyPatch modifies template in-place to be a server-side apply patch.
+func prepareClaimApplyPatch(claim *v1.PersistentVolumeClaim, template *v1.PersistentVolumeClaim) {
+	req := template.Spec.Resources.Requests["storage"]
+	minSize := claim.Spec.Resources.Requests["storage"]
+	if utilfeature.DefaultFeatureGate.Enabled(features.RecoverVolumeExpansionFailure) {
+		minSize = claim.Status.Capacity["storage"]
+	}
+	if req.Cmp(minSize) <= 0 {
+		// We do not shrink PVCs
+		template.Spec.Resources.Requests["storage"] = minSize
+	}
+	// We depends on the old size to set the new size, so add ResourceVersion to avoid conflict.
+	template.ResourceVersion = claim.ResourceVersion
+}
```
I know there's likely some subtleties here that I'm overlooking, but I really like that idea! I.e. of not trying to manipulate the storage requests and just checking readiness.

eakman-datadog replied:
Thought about this a little bit.

If you were to make the change your suggesting of just not shrinking the PVC spec then recovering from a failed expansion would require direct manipulation of the underlying PVCs right? 

If so, that might be okay, but I think it would be useful if there was an explanation of the operational steps required to recover from a failed expansion perhaps in the KEP. Will add a comment there.

Another question I had was, if you look at the item 3 in the proposal [section](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1790-recover-resize-failure/README.md#proposal) of KEP 1790, they proposed adding a `ResizeStatus` field to the PVC API. Could we use this field to more appropriately accommodate the failed expansion recovery use case?
